// @astrojs/ts-plugin@1.8.0/dist/language.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AstroVirtualCode = exports.getLanguageModule = void 0;
const language_core_1 = require("@volar/language-core");
const astro2tsx_js_1 = require("./astro2tsx.js");
function getLanguageModule() {
    return {
        getLanguageId(scriptId) {
            if (scriptId.endsWith('.astro')) {
                return 'astro';
            }
        },
        createVirtualCode(scriptId, languageId, snapshot) {
            if (languageId === 'astro') {
                // scriptId will never be a uri in ts plugin
                const fileName = scriptId;
                return new AstroVirtualCode(fileName, snapshot);
            }
        },
        updateVirtualCode(_scriptId, astroFile, snapshot) {
            astroFile.update(snapshot);
            return astroFile;
        },
        typescript: {
            extraFileExtensions: [{ extension: 'astro', isMixedContent: true, scriptKind: 7 }],
            getServiceScript(astroCode) {
                for (const code of (0, language_core_1.forEachEmbeddedCode)(astroCode)) {
                    if (code.id === 'tsx') {
                        return {
                            code,
                            extension: '.tsx',
                            scriptKind: 4,
                        };
                    }
                }
            },
        },
    };
}
exports.getLanguageModule = getLanguageModule;
class AstroVirtualCode {
    constructor(fileName, snapshot) {
        this.fileName = fileName;
        this.snapshot = snapshot;
        this.id = 'root';
        this.languageId = 'astro';
        this.codegenStacks = [];
        this.onSnapshotUpdated();
    }
    update(newSnapshot) {
        this.snapshot = newSnapshot;
        this.onSnapshotUpdated();
    }
    onSnapshotUpdated() {
        this.mappings = [
            {
                sourceOffsets: [0],
                generatedOffsets: [0],
                lengths: [this.snapshot.getLength()],
                data: {
                    verification: true,
                    completion: true,
                    semantic: true,
                    navigation: true,
                    structure: true,
                    format: false,
                },
            },
        ];
        this.embeddedCodes = [];
        const tsx = (0, astro2tsx_js_1.astro2tsx)(this.snapshot.getText(0, this.snapshot.getLength()), this.fileName);
        this.embeddedCodes.push(tsx.virtualFile);
    }
}
exports.AstroVirtualCode = AstroVirtualCode;
